include/grpc++/impl/codegen/thrift_serializer.h

```diff
@@ -0,0 +1,148 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+ #ifndef GRPCXX_IMPL_CODEGEN_THRIFT_SERIALIZER_H
+ #define GRPCXX_IMPL_CODEGEN_THRIFT_SERIALIZER_H
+
+#include <memory>
+#include <string>
+
+#include <thrift/protocol/TBinaryProtocol.h>
+#include <thrift/protocol/TCompactProtocol.h>
+#include <thrift/transport/TBufferTransports.h>
+#include <thrift/transport/TTransportUtils.h>
+#include <grpc/impl/codegen/byte_buffer.h>
+
+namespace apache {
+namespace thrift {
+namespace util {
+
+using apache::thrift::protocol::TBinaryProtocolT;
+using apache::thrift::protocol::TCompactProtocolT;
+using apache::thrift::protocol::TNetworkBigEndian;
+using apache::thrift::transport::TMemoryBuffer;
+using apache::thrift::transport::TBufferBase;
+using apache::thrift::transport::TTransport;
+using std::shared_ptr;
+
+
+template <typename Dummy, typename P>
+class ThriftSerializer {
+public:
+  ThriftSerializer()
+  : prepared_ (false)
+  , lastDeserialized_ (false)
+  , serializeVersion_ (false) {}
+
+  /**
+   * Serialize the passed type into the internal buffer
+   * and returns a pointer to internal buffer and its size
+   *
+   */
+  template <typename T>
+  void serialize(const T& fields, const uint8_t** serializedBuffer,
+      size_t* serializedLen);
+
+  /**
+   * Serialize the passed type into the byte buffer
+   */
+  template <typename T>
+  void serialize(const T& fields, grpc_byte_buffer** bp);
+
+  /**
+   * Deserialize the passed char array into  the passed type, returns the number
+   * of bytes that have been consumed from the passed string.
+   */
+  template <typename T>
+  uint32_t deserialize(const uint8_t* serializedBuffer, size_t length,
+      T* fields);
+
+  /**
+   * Deserialize the passed byte buffer to passed type, returns the number
+   * of bytes consumed from byte buffer
+   */
+  template <typename T>
+  uint32_t deserialize(grpc_byte_buffer* buffer, T* msg);
+
+  void setSerializeVersion(bool value);
+
+  virtual ~ThriftSerializer() {}
+
+
+  /**
+   * Set the container size limit to deserialize
+   * This function should be called after buffer_ is initialized
+   */
+  void setContainerSizeLimit(int32_t container_limit) {
+    if (!prepared_) {
+      prepare();
+    }
+    protocol_->setContainerSizeLimit(container_limit);
+  }
+
+  /**
+   * Set the string size limit to deserialize
+   * This function should be called after buffer_ is initialized
+   */
+  void setStringSizeLimit(int32_t string_limit) {
+    if (!prepared_) {
+      prepare();
+    }
+    protocol_->setStringSizeLimit(string_limit);
+  }
+
+
+  private:
+    void prepare();
+
+  private:
+    typedef P Protocol;
+    bool prepared_;
+    bool lastDeserialized_;
```

Comment:
C/C++ variable names should be snake_case (so, of course, class member variable names would be snake_case_) .